Content.Server/Shuttles/EntitySystems/ThrusterSystem.cs

```diff
@@ -0,0 +1,485 @@
+        {
+            if (args.Anchored && CanEnable(uid, component))
+            {
+                EnableThruster(uid, component);
+            }
+            else
+            {
+                DisableThruster(uid, component);
+            }
+        }
+
+        private void OnThrusterInit(EntityUid uid, ThrusterComponent component, ComponentInit args)
+        {
+            _ambient.SetAmbience(uid, false);
+
+            if (!component.EnabledVV)
+            {
+                return;
+            }
+
+            if (CanEnable(uid, component))
+            {
+                EnableThruster(uid, component);
+            }
+        }
+
+        private void OnThrusterShutdown(EntityUid uid, ThrusterComponent component, ComponentShutdown args)
+        {
+            DisableThruster(uid, component);
+        }
+
+        private void OnPowerChange(EntityUid uid, ThrusterComponent component, PowerChangedEvent args)
+        {
+            if (args.Powered && CanEnable(uid, component))
+            {
+                EnableThruster(uid, component);
+            }
+            else
+            {
+                DisableThruster(uid, component);
+            }
+        }
+
+        /// <summary>
+        /// Tries to enable the thruster and turn it on. If it's already enabled it does nothing.
+        /// </summary>
+        public void EnableThruster(EntityUid uid, ThrusterComponent component, TransformComponent? xform = null)
+        {
+            if (component.Enabled ||
+                !Resolve(uid, ref xform) ||
+                !_mapManager.TryGetGrid(xform.GridID, out var grid)) return;
+
+            component.Enabled = true;
+
+            if (!EntityManager.TryGetComponent(grid.GridEntityId, out ShuttleComponent? shuttleComponent)) return;
+
+            Logger.DebugS("thruster", $"Enabled thruster {uid}");
+
+            switch (component.Type)
+            {
+                case ThrusterType.Linear:
+                    var direction = (int) xform.LocalRotation.GetCardinalDir() / 2;
+
+                    shuttleComponent.LinearThrusterImpulse[direction] += component.Impulse;
+                    DebugTools.Assert(!shuttleComponent.LinearThrusters[direction].Contains(component));
+                    shuttleComponent.LinearThrusters[direction].Add(component);
+
+                    // Don't just add / remove the fixture whenever the thruster fires because perf
+                    if (EntityManager.TryGetComponent(uid, out PhysicsComponent? physicsComponent))
+                    {
+                        var shape = new PolygonShape();
+                        Span<Vector2> verts = stackalloc Vector2[4];
+
+                        verts[0] = new Vector2(-0.4f, 0.5f);
+                        verts[1] = new Vector2(-0.1f, 1.2f);
+                        verts[2] = new Vector2(0.1f, 1.2f);
+                        verts[3] = new Vector2(0.4f, 0.5f);
```

Comment:
sus that these are hardcoded

Reply:
I mean I guess I could make it a datafield